File: python/lib/asterisk_rest_api.py
"""
 Copyright (C) 2013 Digium, Inc.

 Erin Spiceland <espiceland@digium.com>

 See http://www.asterisk.org for more information about
 the Asterisk project. Please do not directly contact
 any of the maintainers of this project for assistance;
 the project provides a web site, mailing lists and IRC
 channels for your use.

 This program is free software, distributed under the terms
 detailed in the the LICENSE file at the top of the source tree.

"""
import requests
import json
import re
from errors import *


class AsteriskRestAPI:
    def __init__(self, uri='localhost'):
        """Initiate new AsteriskRestAPI instance.

        Takes REST API URI. Default http://localhost:8088/stasis/api
        from AsteriskPy._stasis_base

        """
        self._base_uri = uri

        try:
            requests.get("%s/asterisk.json" % (self._base_uri))
        except requests.exceptions.ConnectionError:
            raise AsteriskPyAccessException(
                "Cannot access URI %s" % (self._base_uri)
            )

    def call(self, object_path,
             http_method='GET', api_method=None,
             parameters=None, object_id=None):
        """Call an Asterisk API method, return result dictionary

        Throws AsteriskPyAccessException if the server is unreachable.
        Returns a dict of the following structure:

        {
            'success' : True, # or False
            'response' : jsonObject, # or None
            'error' : None, # or string
        }

        success indicates the success or failure of the Asterisk API call.
        response is a dictionary constructed by json.dumps(json_string)
        error is a message.

        If the API call is successful but Asterisk returns invalid JSON, error
        will be "Invalid JSON." and response will be the unchanged content
        of the response.

        """
        result = {'success' : False, 'response' : None, 'error' : None}
        if object_id:
            object_path = object_path % (object_id)

        request_uri = "%s/%s" % (self._base_uri, object_path)
        if object_id is not None:
            request_uri = request_uri + "/%s" % (object_id)

        if api_method is not None:
            request_uri = request_uri + "/%s" % (api_method)

        try:
            if http_method == 'GET':
                resp = requests.get(request_uri, params=parameters)
            elif http_method == 'POST':
                resp = requests.post(request_uri, params=parameters)
            elif http_method == 'DELETE':
                resp = requests.delete(request_uri, params=parameters)
            elif http_method == 'PUT':
                resp = requests.put(request_uri, params=parameters)
        except requests.exceptions.ConnectionError:
            raise self.AsteriskPyAccessException(
                "Cannot access URI %s" % (request_uri)
            )

        if resp is None:
            # No response or exception? This will probably never happen.
            result['error'] = "No response."
            return result

        if resp.status_code in [418, 200]:
            result['success'] = True
        else:
            result['error'] = "HTTP error occurred: %s" % (resp.status_code)
            return result

        try:
            result['response'] = json.loads(resp.text)
        except ValueError:
            result['response'] = resp.text
            result['error'] = "Invalid JSON."

        return result

    def add_event_handler(self, event_name, handler):
        """Add a general event handler for Stasis events.
        For object-specific events, use the object's add_event_handler instead.
        """
        pass

    def remove_event_handler(self, event_name, handler):
        """Add a general event handler for Stasis events.
        For object-specific events, use the object's add_event_handler instead.
        """
        pass

    class AsteriskPyAccessException(Exception):
        def __init__(self, m):
            self.message = m

        def __str__(self):
            return self.message

File: javascript/lib/asterisk_rest_api.js
/*****************************************************************************
* Copyright (C) 2013 Digium, Inc.
* All Rights Reserved.
******************************************************************************
* @Package: AsteriskRESTAPI
* @Authors: Erin Spiceland <espiceland@digium.com>
*
* See http://www.asterisk.org for more information about
* the Asterisk project. Please do not directly contact
* any of the maintainers of this project for assistance;
* the project provides a web site, mailing lists and IRC
* channels for your use.
*
* This program is free software, distributed under the terms
* detailed in the the LICENSE file at the top of the source tree.
*
*****************************************************************************/
AsteriskJs.AsteriskRESTAPI = function(parameters) {
	/* Handle HTTP requests to Asterisk */
	var this_ = this;
	for (var key in parameters) {
		this[key] = parameters[key];
	}

	if (!this.stasis_base) {
		throw new Error("Can't call Asterisk REST API without stasis_base.");
	}

	if (!this.stasis_base.match(/https*:\/\/.+\/stasis\/*/i)) {
		throw new Error("stasis_base value is invalid: " + this.stasis_base);
	}

	this.call = function(params) {
		/*
		Call an Asterisk API method, call responseHandler with request info
		"data" including URI and parameters.

		responseHandler(response, data);

		success indicates the success or failure of the Asterisk API call.
		response is a dictionary constructed by json.dumps(json_string)
		error is a message.

		If the API call is successful but Asterisk returns invalid JSON, error
		will be "Invalid JSON." and response will be the unchanged content
		of the response.
		*/

		if (!this.hasOwnProperty('responseHandler')) {
			this.responseHandler = function(response, data) {
				console.log('default responseHandler');
				console.log(response);
			};
		}

		if (!params) {
			throw new Error('No params.');
		}

		if (!params.http_method) {
			throw new Error('No HTTP method.');
		}


		if (params.object_id) {
			params.path = params.path.replace(/\%s/ig, params.object_id);
		}

		var uri = this.stasis_base + params.path;
		var response;
		var xhr = new XMLHttpRequest();

		/* Params go in the URI for GET, DELETE, same format for
		 * POST and PUT, but they must be sent separately after the
		 * request is opened. */
		var paramString = make_param_string(params.parameters);
		if (['GET', 'DELETE'].indexOf(params.http_method) > -1) {
			uri += paramString;
			paramString = null;
		}

		xhr.open(params.http_method, uri);
		if (['POST', 'PUT'].indexOf(params.http_method) > -1) {
			try {
				xhr.setRequestHeader(
					'Content-type',
					'application/x-www-form-urlencoded'
				);
			} catch (e) {
				console.log("Can't set content-type header in readyState "
					+ xhr.readyState + ". " + e.message);
			}
			xhr.setRequestHeader('Content-length', params.length);
			xhr.setRequestHeader('Connection', 'close');
		} else if (['GET', 'DELETE'].indexOf(params.http_method) == -1) {
			throw new Error('Illegal HTTP request method '
				+ params.http_method);
		}

		try {
			xhr.send(paramString);
		} catch (e) {
			console.log("Can't call xhr.send. " + e.message);
		}
		xhr.onreadystatechange = function() {
			if (this.readyState !== 4) { return; }
			if (this.status === 0) { return; }
			/*console.log('readyState is ' + this.readyState + " and status is "
				+ this.status + " and text is " + this.statusText);*/
			if ([200, 204, 205, 302, 404, 418].indexOf(this.status) > -1) {
				this_.responseHandler(this, {
					'uri' : uri,
					'params' : params.parameters
				});
			} else {
				console.log('unexpected response ' + this.status);
			}
		};
	};

	var make_param_string = function(params) {
		/* Turn key/value and key/list pairs into an HTTP URL parameter string.
		var1=value1&var2=value2,value3,value4 */
		var strings = [];
		for (var name in params) { if (params.hasOwnProperty(name)) {
			/* Skip objects -- We won't know how to name these. */
			if (typeof params[name] == 'array') {
				strings.push([name, params[name].join(',')].join('='));
			} else if (typeof params[name] != 'object'
					&& typeof params[name] != 'function') {
				strings.push([name, params[name]].join('='));
			}
		}}
		if (strings.length > 0) {
			return '?' + strings.join('&');
		} else {
			return '';
		}
	};
};
